virtual_rainforest/models/hydrology/hydrology_model.py

```diff
@@ -204,242 +222,228 @@ def spinup(self) -> None:
+                0,
+                self.constants.soil_moisture_capacity,
+            ),
+        )
 
-    # Expand to all soil layers and add atmospheric layers (nan)
-    output["soil_moisture"] = xr.concat(
-        [
-            DataArray(
-                np.full(
-                    (
-                        len(layer_roles) - layer_roles.count("soil"),
-                        len(layer_heights.cell_id),
+        # Expand to all soil layers and add atmospheric layers (nan)
+        soil_hydrology["soil_moisture"] = xr.concat(
+            [
+                DataArray(
+                    np.full(
+                        (
+                            len(self.layer_roles) - self.layer_roles.count("soil"),
+                            len(self.data["layer_heights"].cell_id),
+                        ),
+                        np.nan,
                     ),
-                    np.nan,
+                    dims=["layers", "cell_id"],
                 ),
-                dims=["layers", "cell_id"],
-            ),
-            soil_moisture_reduced.expand_dims(
-                dim={"layers": layer_roles.count("soil")},
-            ),
-        ],
-        dim="layers",
-    ).assign_coords(
-        coords={
-            "layers": np.arange(len(layer_roles)),
-            "layer_roles": ("layers", layer_roles),
-            "cell_id": layer_heights.cell_id,
-        }
-    )
+                soil_moisture_reduced.expand_dims(
+                    dim={"layers": self.layer_roles.count("soil")},
+                ),
+            ],
+            dim="layers",
+        ).assign_coords(
+            coords={
+                "layers": np.arange(len(self.layer_roles)),
+                "layer_roles": ("layers", self.layer_roles),
+                "cell_id": self.data["layer_heights"].cell_id,
+            }
+        )
 
-    return output
+        # Update data object
+        self.data.add_from_dict(output_dict=soil_hydrology)
+
+    def cleanup(self) -> None:
+        """Placeholder function for hydrology model cleanup."""
 
 
 def calculate_vertical_flow(
     soil_moisture: DataArray,
     soil_depth: DataArray,
-    soil_moisture_capacity: Union[float, DataArray] = (
-        HydroConsts["soil_moisture_capacity"]
-    ),
-    soil_moisture_residual: Union[float, DataArray] = (
-        HydroConsts["soil_moisture_residual"]
-    ),
-    hydraulic_conductivity: Union[float, DataArray] = (
-        HydroConsts["hydraulic_conductivity"]
-    ),
-    hydraulic_gradient: Union[float, DataArray] = (HydroConsts["hydraulic_gradient"]),
-    timestep_conversion_factor: float = (HydroConsts["seconds_to_month"]),
-    nonlinearily_parameter: Union[float, DataArray] = (
-        HydroConsts["nonlinearily_parameter"]
-    ),
-    meters_to_millimeters: float = HydroConsts["meters_to_millimeters"],
+    soil_moisture_capacity: Union[float, DataArray],
+    soil_moisture_residual: Union[float, DataArray],
```

Comment:
Are these `Union`'s because you can either supply them with real data or just a constant fixed across all grid cells?

Reply:
Actually - this is a thing I went through in `pyrealm` and asked @dalonsoa and @alexdewar about. Having this `Union` typing for scalar versus arrays inputs makes things really long-winded, compared to simply asserting that it is an array (which could be a scalar array).

In my case, I was dealing with scalars and np.arrays and had lots of Union statements, but I ended up ripping all of that out and just asserting that all things were going to be arrays. I'm not sure it works quite so cleanly with DataArrays, but I think it is worth thinking about insisting on one kind of input and simplifying this?

Reply:
I had a look into it and the problem with `DataArray` seems to be that the dimensions have to match, so the constants have to have at least the `cell_id `dimension. It seems to me like more of a clutter to introduce that than having the `Union` statement.